### PR TITLE
[VL] Validate unsupported regex in re2

### PR DESCRIPTION
**Pull request by marin-ma**

### CODE REVIEW

File: cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc [hunk 1]
```diff
@@ -150,9 +151,18 @@ bool SubstraitToVeloxPlanValidator::validateScalarFunction(
   if (name == "extract") {
     return validateExtractExpr(params);
   }
-  if (name == "regexp_extract_all" && !scalarFunction.arguments()[1].value().has_literal()) {
-    logValidateMsg("native validation failed due to: pattern is not literal for regex_extract_all.");
-    return false;
+  if (name == "regexp_extract_all" || name == "regexp_extract" || name == "regexp_replace" || name == "rlike") {
+    auto patternArg = scalarFunction.arguments()[1].value();
+    if (!patternArg.has_literal()) {
+      logValidateMsg("native validation failed due to: pattern is not literal for regex_extract_all.");
```
Comment (PHILO-HE) [inline cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc:157]: Nit：
We should use `name` to replace `regex_extract_all`. Thanks!

**Review by PHILO-HE — COMMENTED** [review]
Looks great!
Could you add some scala tests? e.g, with unsupported lookaround pattern used.

File: cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc [hunk 2]
```diff
@@ -150,9 +151,18 @@ bool SubstraitToVeloxPlanValidator::validateScalarFunction(
...
+      return false;
+    }
+    auto pattern = patternArg.literal().string();
+    auto re2 = RE2(pattern, RE2::Quiet);
+    if (!re2.ok()) {
+      logValidateMsg("native validation failed due to: pattern " + pattern + " compilation failed in RE2.");
```
Comment (PHILO-HE) [inline cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc:163]: Can we expose RE2's compilation error message to the validation message? It may be helpful to users to figure out the root reason. Thanks!

File: cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc [hunk 3]
```diff
@@ -150,9 +151,18 @@ bool SubstraitToVeloxPlanValidator::validateScalarFunction(
...
-    logValidateMsg("native validation failed due to: pattern is not literal for regex_extract_all.");
-    return false;
+  if (name == "regexp_extract_all" || name == "regexp_extract" || name == "regexp_replace" || name == "rlike") {
+    auto patternArg = scalarFunction.arguments()[1].value();
+    if (!patternArg.has_literal()) {
+      logValidateMsg("native validation failed due to: pattern is not literal for regex_extract_all.");
+      return false;
+    }
+    auto pattern = patternArg.literal().string();
+    auto re2 = RE2(pattern, RE2::Quiet);
```
Comment (unigof) [inline cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc:161]: Could you extract **re2** related code to utils? Maybe it will be used frequently.

File: cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc (hunk 2 above)
Comment (marin-ma) [reply cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc:163]: Weird but the RE2 error message is always empty.

Comment (PHILO-HE) [reply cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc:163]: Maybe, `Quiet` option makes it silent. I'm not sure. You can try to change it. Thanks!

Comment (marin-ma) [reply cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc:163]: Tried default options with log print but not help. I'm not sure if it's an issue from the current RE2 version we use.

File: cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc (hunk 3 above)
Comment (marin-ma) [reply cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc:161]: Sorry for missing this. Will do.

File: cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc (hunk 2 above)
Comment (marin-ma) [reply cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc]: Just found it's due to my local env was broken. Already added.

File: cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc [hunk 4]
```diff
@@ -150,9 +152,18 @@ bool SubstraitToVeloxPlanValidator::validateScalarFunction(
   if (name == "extract") {
     return validateExtractExpr(params);
   }
-  if (name == "regexp_extract_all" && !scalarFunction.arguments()[1].value().has_literal()) {
-    logValidateMsg("native validation failed due to: pattern is not literal for regex_extract_all.");
-    return false;
+  if (name == "regexp_extract_all" || name == "regexp_extract" || name == "regexp_replace" || name == "rlike") {
```
Comment (jackylee-ch) [inline cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc:155]: nit: There are 10 spark build-in functions related to regex, it is better to use `list.contains` to define whether to check pattern.

Comment (marin-ma) [reply cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc:155]: Could you provide those function names or hint me where to find them? I only found these 4 functions.

Comment (jackylee-ch) [reply cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc:155]: We are making split to support regex, so split also needs to be added here later.

Comment (marin-ma) [reply cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc:155]: > Here is the list for functions that starts with regexp. https://spark.apache.org/docs/latest/api/sql/index.html#regexp And, `rlike` and `split` use regex too.

The url points to the latest spark function list. Many functions are not supported in spark 3.3.2 https://spark.apache.org/docs/3.3.2/api/sql/index.html#regexp_extract

Currently there should still be the 4 functions. `regex`, `regex_like` and `rlike` are the same and all aliased to `rlike.
`split` is in blacklist. I suppose there's no missing functions.

Comment (jackylee-ch) [reply cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc]: Okey, we can change it after `split` support regex with re2.

File: cpp/velox/utils/Common.cc [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Common.h"
+
+namespace gluten {
+
+std::unique_ptr<re2::RE2> compilePattern(const std::string& pattern) {
+  return std::make_unique<re2::RE2>(re2::StringPiece(pattern), RE2::Quiet);
```
Comment (jackylee-ch) [inline cpp/velox/utils/Common.cc:23]: nit: add a check whether the pattern semantic diff between re2 and java.util.regex. Such as https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/RegexFunctions.cpp#L26

Comment (unigof) [reply cpp/velox/utils/Common.cc:23]: Hi, @jackylee-ch ,The check for regex semantic diff on [RegexFunctions.cpp#L26] maybe not enough, do you have any idea for diff check?
These are some other diff case：
'|' 
'.'
'.*'
'\\.*'
'\\?*'

File: cpp/velox/utils/Common.cc [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+#include "Common.h"
+
+namespace gluten {
+
+// Note: This method is mostly copied from velox/functions/sparksql/RegexFunctions.cpp
+//
+// Blocks patterns that contain character class union, intersection, or
+// difference because these are not understood by RE2 and will be parsed as a
+// different pattern than in java.util.regex.
+bool ensureRegexIsCompatible(const std::string& pattern, std::string& error) {
```
Comment (marin-ma) [inline cpp/velox/utils/Common.cc:27]: @jackylee-ch Added here. @unigof As for your concern, could you submit another PR to enhance this function after this PR merge?

**Review by PHILO-HE — APPROVED** [review]
Looks great! This fix is much better!
